File: stock-analyzer/backend/analysis/social.py
"""Reddit/WallStreetBets sentiment: mention counts, VADER sentiment, WSB meme signals, and mention trend.

Data source: Reddit API via ``praw``. Falls back to a "credentials not
configured" shape until a user sets up ``backend/.env``.
"""

import os
import re
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional

import praw
from dotenv import load_dotenv

from utils.sentiment_utils import aggregate_sentiment, score_items

load_dotenv()

SUBREDDITS = ["wallstreetbets", "investing", "stocks", "SecurityAnalysis", "options"]
YOLO_KEYWORDS = ["YOLO", "CALLS", "PUTS", "TENDIES"]

DEFAULT_MENTION_SEARCH_LIMIT = 100
WSB_HOT_POST_LIMIT = 50
WSB_SEARCH_LIMIT = 50

# --- Mention trend ---
MENTION_TREND_DAYS = 7
TRENDING_MULTIPLIER = 2
TREND_WINDOW_DAYS = 3

# --- Combined social signal ---
SOCIAL_SIGNAL_BULLISH_MIN = 1
SOCIAL_SIGNAL_BEARISH_MAX = -1


class SocialSentimentAnalyzer:
    """Fetches and scores Reddit/WSB sentiment for one ticker."""

    def __init__(self, ticker: str):
        self.ticker = ticker.upper()
        self._reddit = None
        self._reddit_checked = False
        self._mentions_cache = None

    def setup_reddit(self) -> Optional["praw.Reddit"]:
        """Lazily-initialized, cached read-only Reddit client, or None if unconfigured."""
        if self._reddit_checked:
            return self._reddit
        self._reddit_checked = True

        client_id = os.getenv("REDDIT_CLIENT_ID")
        client_secret = os.getenv("REDDIT_CLIENT_SECRET")
        user_agent = os.getenv("REDDIT_USER_AGENT", "StockAnalyzer/1.0")

        placeholder = (
            not client_id or not client_secret
            or client_id.startswith("your_") or client_secret.startswith("your_")
        )
        if placeholder:
            self._reddit = None
            return None

        try:
            reddit = praw.Reddit(client_id=client_id, client_secret=client_secret, user_agent=user_agent)
            reddit.read_only = True
            self._reddit = reddit
        except Exception:
            self._reddit = None

        return self._reddit

    def get_reddit_mentions(self, limit: int = DEFAULT_MENTION_SEARCH_LIMIT) -> Optional[Dict]:
        """Search all `SUBREDDITS` for `$TICKER`/`TICKER` mentions, deduped by post ID and cached."""
        if self._mentions_cache is not None:
            return self._mentions_cache

        reddit = self.setup_reddit()
        if reddit is None:
            return None

        posts = {}
        mentions_by_subreddit = {sub: 0 for sub in SUBREDDITS}
        queries = [f"${self.ticker}", self.ticker]

        try:
            for sub_name in SUBREDDITS:
                subreddit = reddit.subreddit(sub_name)
                for query in queries:
                    try:
                        for submission in subreddit.search(query, limit=limit, sort="new"):
                            if submission.id not in posts:
                                posts[submission.id] = {
                                    "id": submission.id,
                                    "title": submission.title,
                                    "score": submission.score,
                                    "num_comments": submission.num_comments,
                                    "created_utc": submission.created_utc,
                                    "url": submission.url,
                                    "subreddit": sub_name,
                                }
                                mentions_by_subreddit[sub_name] += 1
                    except Exception:
                        continue

            result = {
                "total_mentions": len(posts),
                "posts_found": list(posts.values()),
                "mentions_by_subreddit": mentions_by_subreddit,
            }
        except Exception:
            result = {"total_mentions": 0, "posts_found": [], "mentions_by_subreddit": mentions_by_subreddit}

        self._mentions_cache = result
        return result

    def get_reddit_sentiment(self) -> Optional[Dict]:
        """VADER sentiment across all mentioned posts' titles, or None if unavailable."""
        mentions = self.get_reddit_mentions()
        if mentions is None:
            return None

        try:
            scored_posts = score_items(mentions["posts_found"])
            aggregate = aggregate_sentiment(scored_posts)
            return {
                "avg_compound": aggregate["avg_compound"],
                "counts": aggregate["counts"],
                "overall_reddit_sentiment": aggregate["overall_sentiment"],
                "upvote_weighted_sentiment": aggregate["upvote_weighted_sentiment"],
                "posts": aggregate["posts"],
            }
        except Exception:
            return None

    def _wsb_signals(self, ticker_posts: List[dict]) -> Dict:
        """Rocket-emoji count, YOLO/loss-porn keyword mentions, and DD posts for a post list."""
        titles_upper = [p["title"].upper() for p in ticker_posts]
        rocket_count = sum(title.count("\U0001F680") for title in titles_upper)
        yolo_mentions = sum(1 for title in titles_upper if any(kw in title for kw in YOLO_KEYWORDS))
        loss_porn = sum(1 for title in titles_upper if "LOSS" in title or "RIP" in title)
        dd_posts = [
            p for p, title in zip(ticker_posts, titles_upper)
            if re.search(r"\bDD\b", title) or "DUE DILIGENCE" in title
        ]
        return {"rocket_count": rocket_count, "yolo_mentions": yolo_mentions, "loss_porn": loss_porn, "dd_posts": dd_posts}

    def get_wsb_analysis(self) -> Optional[Dict]:
        """WallStreetBets mentions, sentiment, and meme signals for this ticker.

        Combines current hot posts with a direct search, since a trending
        ticker might not match a literal title search or vice versa.
        """
        reddit = self.setup_reddit()
        if reddit is None:
            return None

        try:
            subreddit = reddit.subreddit("wallstreetbets")
            posts = {}

            try:
                for submission in subreddit.hot(limit=WSB_HOT_POST_LIMIT):
                    posts[submission.id] = {
                        "id": submission.id, "title": submission.title, "score": submission.score,
                        "num_comments": submission.num_comments, "created_utc": submission.created_utc,
                        "url": submission.url, "subreddit": "wallstreetbets",
                    }
            except Exception:
                pass

            try:
                for submission in subreddit.search(self.ticker, limit=WSB_SEARCH_LIMIT, sort="new"):
                    posts[submission.id] = {
                        "id": submission.id, "title": submission.title, "score": submission.score,
                        "num_comments": submission.num_comments, "created_utc": submission.created_utc,
                        "url": submission.url, "subreddit": "wallstreetbets",
                    }
            except Exception:
                pass

            ticker_upper = self.ticker.upper()
            ticker_posts = [
                p for p in posts.values()
                if ticker_upper in p["title"].upper() or f"${ticker_upper}" in p["title"].upper()
            ]

            wsb_signals = self._wsb_signals(ticker_posts)
            scored_posts = score_items(ticker_posts)
            aggregate = aggregate_sentiment(scored_posts)

            return {
                "wsb_mentions": len(ticker_posts),
                "wsb_sentiment": {
                    "avg_compound": aggregate["avg_compound"],
                    "counts": aggregate["counts"],
                    "overall_sentiment": aggregate["overall_sentiment"],
                },
                "wsb_signals": {
                    "rocket_count": wsb_signals["rocket_count"],
                    "yolo_mentions": wsb_signals["yolo_mentions"],
                    "loss_porn": wsb_signals["loss_porn"],
                },
                "dd_posts": wsb_signals["dd_posts"],
            }
        except Exception:
            return None

    def get_mention_trend(self) -> Optional[Dict]:
        """Daily mention counts over the last `MENTION_TREND_DAYS` days, plus a trend direction."""
        mentions = self.get_reddit_mentions()
        if mentions is None:
            return None

        try:
            now = datetime.now(timezone.utc)
            days = [(now - timedelta(days=i)).date() for i in range(MENTION_TREND_DAYS - 1, -1, -1)]
            daily_mentions = {str(d): 0 for d in days}

            for post in mentions["posts_found"]:
                post_date = datetime.fromtimestamp(post["created_utc"], tz=timezone.utc).date()
                key = str(post_date)
                if key in daily_mentions:
                    daily_mentions[key] += 1

            counts = list(daily_mentions.values())
            today_count = counts[-1]
            avg_count = sum(counts) / len(counts) if counts else 0
            is_trending = bool(avg_count and today_count > TRENDING_MULTIPLIER * avg_count)

            first_half_avg = sum(counts[:TREND_WINDOW_DAYS]) / TREND_WINDOW_DAYS
            second_half_avg = sum(counts[-TREND_WINDOW_DAYS:]) / TREND_WINDOW_DAYS
            if second_half_avg > first_half_avg:
                trend_direction = "increasing"
            elif second_half_avg < first_half_avg:
                trend_direction = "decreasing"
            else:
                trend_direction = "stable"

            return {
                "daily_mentions": daily_mentions,
                "is_trending": is_trending,
                "trend_direction": trend_direction,
            }
        except Exception:
            return None

    def get_future_features(self) -> Dict:
        """Static roadmap block describing planned X/Twitter and YouTube sentiment features."""
        ticker = self.ticker
        return {
            "twitter_x": {
                "status": "coming_soon",
                "description": "Real-time X/Twitter sentiment using cashtags ($TICKER)",
                "planned_features": [
                    "Live cashtag feed via X API v2",
                    "Sentiment analysis on tweets",
                    "Key financial accounts monitoring",
                    "Tweet volume trends",
                ],
                "links": {
                    "live_feed": f"https://x.com/search?q={ticker}&f=live",
                    "top_posts": f"https://x.com/search?q={ticker}+stock&f=top",
                    "cashtag": f"https://x.com/search?q=%24{ticker}&src=cashtag_click",
                },
                "api_docs": "https://developer.x.com/en/docs/x-api",
            },
            "youtube": {
                "status": "coming_soon",
                "description": "YouTube financial video analysis for the ticker",
                "planned_features": [
                    "Latest videos from top financial YouTubers",
                    "View count and engagement metrics",
                    "Title sentiment analysis via VADER",
                    "Filtered by trusted channels: Graham Stephan, Meet Kevin, ZipTrader, etc.",
                ],
                "api_docs": "https://developers.google.com/youtube/v3",
            },
        }

    def _social_signal(self, reddit_sentiment: Optional[Dict], mention_trend: Optional[Dict], wsb: Optional[Dict]) -> str:
        """Combine Reddit sentiment, a mention-trend spike, and WSB loss-porn/rocket balance."""
        score = 0
        overall_sentiment = reddit_sentiment.get("overall_reddit_sentiment") if reddit_sentiment else None
        if overall_sentiment in ("Bullish", "Very Bullish"):
            score += 1
        elif overall_sentiment in ("Bearish", "Very Bearish"):
            score -= 1

        if mention_trend and mention_trend.get("is_trending"):
            score += 1

        if wsb:
            signals = wsb.get("wsb_signals", {})
            if signals.get("loss_porn", 0) > signals.get("rocket_count", 0):
                score -= 1

        if score >= SOCIAL_SIGNAL_BULLISH_MIN:
            return "Bullish"
        if score <= SOCIAL_SIGNAL_BEARISH_MAX:
            return "Bearish"
        return "Neutral"

    def get_full_social_analysis(self) -> Dict:
        """Aggregate Reddit mentions/sentiment, WSB signals, mention trend, and an overall signal."""
        if self.setup_reddit() is None:
            return {
                "error": "Reddit credentials not configured",
                "instructions": {
                    "step1": "Go to reddit.com/prefs/apps",
                    "step2": "Click 'create app' at the bottom",
                    "step3": "Select type: script",
                    "step4": "Copy client_id (under app name) and client_secret",
                    "step5": "Add to backend/.env file",
                },
                "future_features": self.get_future_features(),
            }

        try:
            reddit_mentions = self.get_reddit_mentions()
        except Exception:
            reddit_mentions = None

        try:
            reddit_sentiment = self.get_reddit_sentiment()
        except Exception:
            reddit_sentiment = None

        try:
            wsb = self.get_wsb_analysis()
        except Exception:
            wsb = None

        try:
            mention_trend = self.get_mention_trend()
        except Exception:
            mention_trend = None

        return {
            "reddit": {
                "mentions": reddit_mentions,
                "sentiment": reddit_sentiment,
            },
            "wsb": wsb,
            "mention_trend": mention_trend,
            "future_features": self.get_future_features(),
            "social_signal": self._social_signal(reddit_sentiment, mention_trend, wsb),
        }

File: stock-analyzer/frontend/src/pages/Social.js
import { BarChart, Bar, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer } from "recharts";
import useFetch from "../hooks/useFetch";
import { fetchSocial } from "../api/stockApi";
import LoadingSpinner from "../components/LoadingSpinner";
import SectionCard from "../components/SectionCard";
import DataTable from "../components/DataTable";
import SignalBadge from "../components/SignalBadge";
import MetricCard from "../components/MetricCard";
import { fmtNum } from "../utils/format";

function StatRow({ label, value }) {
  return (
    <div style={{ display: "flex", justifyContent: "space-between", padding: "7px 0", borderBottom: "1px solid var(--row-border)", fontSize: 14 }}>
      <span style={{ color: "var(--text-secondary)" }}>{label}</span>
      <span style={{ fontWeight: 600 }}>{value}</span>
    </div>
  );
}

function NotConfigured({ ticker, instructions }) {
  return (
    <div style={{ maxWidth: 1200 }}>
      <h1 style={{ fontSize: 22, marginBottom: 20 }}>Social Sentiment — {ticker}</h1>
      <SectionCard title="REDDIT NOT CONNECTED">
        <p style={{ color: "var(--text-secondary)", fontSize: 13, marginBottom: 10 }}>
          Reddit requires free API credentials to fetch live data. Reddit needs to verify who is making requests
          to protect their platform from abuse. Your credentials are private and only used to read public posts.
        </p>
        <ol style={{ paddingLeft: 20, lineHeight: 1.9, fontSize: 14 }}>
          {Object.values(instructions || {}).map((step, i) => <li key={i}>{step}</li>)}
        </ol>
        <p style={{ color: "var(--text-secondary)", fontSize: 13, marginTop: 10 }}>
          Once connected, this page shows live mentions across r/wallstreetbets, r/investing, r/stocks and
          r/SecurityAnalysis with VADER sentiment scoring on every post.
        </p>
      </SectionCard>

      <SectionCard title="TWITTER/X">
        <p style={{ color: "var(--text-secondary)", fontSize: 13, marginBottom: 10 }}>
          Real-time X/Twitter sentiment using cashtags. Requires X API v2 access (paid, starts at $100/month).
        </p>
        <ol style={{ paddingLeft: 20, lineHeight: 1.9, fontSize: 14 }}>
          <li>Go to developer.x.com and create a developer account</li>
          <li>Create a new project and app in the developer portal</li>
          <li>Apply for Basic tier access ($100/month) or higher</li>
          <li>Copy your API Key, API Secret, Bearer Token</li>
          <li>Add to backend/.env: TWITTER_API_KEY=your_key_here</li>
          <li>In backend/analysis/social.py implement get_twitter_mentions() using the tweepy library</li>
          <li>Add GET /stock/{"{ticker}"}/twitter endpoint to main.py</li>
          <li>Restart the backend server</li>
        </ol>
        <ul style={{ paddingLeft: 20, fontSize: 13, color: "var(--text-secondary)", marginTop: 10 }}>
          <li>Live cashtag feed via X API v2</li>
          <li>Sentiment analysis on tweets with VADER</li>
          <li>Key financial accounts monitoring</li>
          <li>Tweet volume trends over time</li>
        </ul>
      </SectionCard>

      <SectionCard title="YOUTUBE">
        <p style={{ color: "var(--text-secondary)", fontSize: 13, marginBottom: 10 }}>
          YouTube financial video analysis for the ticker. Requires a free Google API key (10,000 requests/day free tier).
        </p>
        <ol style={{ paddingLeft: 20, lineHeight: 1.9, fontSize: 14 }}>
          <li>Go to console.cloud.google.com and sign in</li>
          <li>Create a new project (or use existing one)</li>
          <li>Click "Enable APIs and Services"</li>
          <li>Search for "YouTube Data API v3" and enable it</li>
          <li>Go to "Credentials" and click "Create Credentials"</li>
          <li>Select "API Key" and copy the generated key</li>
          <li>Add to backend/.env: YOUTUBE_API_KEY=your_key_here</li>
          <li>Restart the backend server — YouTube data will appear automatically on this page</li>
        </ol>
        <ul style={{ paddingLeft: 20, fontSize: 13, color: "var(--text-secondary)", marginTop: 10 }}>
          <li>Latest videos from top financial YouTubers</li>
          <li>View count and engagement metrics</li>
          <li>Title sentiment analysis via VADER</li>
          <li>Filtered by trusted channels: Graham Stephan, Meet Kevin, ZipTrader, Andrei Jikh, InvestAnswers</li>
        </ul>
      </SectionCard>
    </div>
  );
}

export default function Social({ ticker }) {
  const { data, loading, error } = useFetch(fetchSocial, ticker);

  if (loading) return <LoadingSpinner label={`Loading social sentiment for ${ticker}...`} />;
  if (error) return <p style={{ color: "var(--red)" }}>Error: {error}</p>;
  if (!data) return <p style={{ color: "var(--red)" }}>No social data available.</p>;

  if (data.error) {
    return <NotConfigured ticker={ticker} instructions={data.instructions} />;
  }

  const { reddit = {}, wsb = {}, mention_trend: trend = {}, social_signal } = data;
  const mentions = reddit.mentions || {};
  const sentiment = reddit.sentiment || {};

  const subredditChart = Object.entries(mentions.mentions_by_subreddit || {}).map(([sub, count]) => ({ sub, count }));
  const dailyChart = Object.entries(trend.daily_mentions || {}).map(([date, count]) => ({ date, count }));

  return (
    <div style={{ maxWidth: 1200 }}>
      <h1 style={{ fontSize: 22, marginBottom: 20 }}>Social Sentiment — {ticker}</h1>

      <SectionCard title="Social Signal">
        <SignalBadge signal={social_signal} size="lg" />
      </SectionCard>

      <div style={{ display: "grid", gridTemplateColumns: "repeat(auto-fit, minmax(160px, 1fr))", gap: 12, marginBottom: 20 }}>
        <MetricCard label="Total Mentions" value={mentions.total_mentions ?? "—"} />
        <MetricCard label="Reddit Sentiment" value={sentiment.overall_reddit_sentiment || "—"} />
        <MetricCard label="Trending?" value={trend.is_trending ? "Yes" : "No"} sub={trend.trend_direction} />
      </div>

      <div style={{ display: "grid", gridTemplateColumns: "repeat(auto-fit, minmax(360px, 1fr))", gap: 18 }}>
        <SectionCard title="Mentions by Subreddit">
          <ResponsiveContainer width="100%" height={200}>
            <BarChart data={subredditChart}>
              <CartesianGrid strokeDasharray="3 3" stroke="var(--border)" />
              <XAxis dataKey="sub" tick={{ fontSize: 10, fill: "var(--text-secondary)" }} />
              <YAxis tick={{ fontSize: 11, fill: "var(--text-secondary)" }} allowDecimals={false} />
              <Tooltip contentStyle={{ background: "var(--card)", border: "1px solid var(--border)", borderRadius: 8 }} />
              <Bar dataKey="count" fill="var(--accent)" radius={[4, 4, 0, 0]} />
            </BarChart>
          </ResponsiveContainer>
        </SectionCard>

        <SectionCard title="Daily Mentions (7d)">
          <ResponsiveContainer width="100%" height={200}>
            <BarChart data={dailyChart}>
              <CartesianGrid strokeDasharray="3 3" stroke="var(--border)" />
              <XAxis dataKey="date" tick={{ fontSize: 10, fill: "var(--text-secondary)" }} />
              <YAxis tick={{ fontSize: 11, fill: "var(--text-secondary)" }} allowDecimals={false} />
              <Tooltip contentStyle={{ background: "var(--card)", border: "1px solid var(--border)", borderRadius: 8 }} />
              <Bar dataKey="count" fill="var(--yellow)" radius={[4, 4, 0, 0]} />
            </BarChart>
          </ResponsiveContainer>
        </SectionCard>
      </div>

      <SectionCard title="WallStreetBets Signals">
        <div style={{ display: "grid", gridTemplateColumns: "repeat(auto-fit, minmax(160px, 1fr))", gap: 12 }}>
          <StatRow label="🚀 Rocket Mentions" value={wsb.wsb_signals?.rocket_count ?? "—"} />
          <StatRow label="YOLO Mentions" value={wsb.wsb_signals?.yolo_mentions ?? "—"} />
          <StatRow label="Loss Porn" value={wsb.wsb_signals?.loss_porn ?? "—"} />
          <StatRow label="DD Posts" value={wsb.dd_posts?.length ?? 0} />
        </div>
        <p style={{ marginTop: 12, fontSize: 13, color: "var(--text-secondary)" }}>
          WSB Sentiment: <SignalBadge signal={wsb.wsb_sentiment?.overall_sentiment} size="sm" />
        </p>
      </SectionCard>

      <SectionCard title="Recent Reddit Posts">
        <DataTable
          columns={[
            { key: "subreddit", label: "Subreddit" },
            { key: "title", label: "Title", render: (v, row) => (row.url ? <a href={row.url} target="_blank" rel="noreferrer" style={{ color: "var(--text)" }}>{v}</a> : v) },
            { key: "score", label: "Upvotes", render: (v) => fmtNum(v, 0) },
            { key: "sentiment_label", label: "Sentiment", render: (v) => (v ? <SignalBadge signal={v} size="sm" /> : "—") },
          ]}
          rows={sentiment.posts}
          emptyText="No Reddit posts found."
        />
      </SectionCard>
    </div>
  );
}
